### PR TITLE
OpenMPTarget team_policy reduce fixes for init/join reductions

### PR DESCRIPTION
**Pull request by rgayatri23**
The PR fixes the compile time issues arising in kokkos-kernels with OpenMPTarget backend. #4500

### CODE REVIEW

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp [hunk 1]
```diff
@@ -454,13 +455,16 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
   void execute() const {
     if constexpr (HasJoin) {
       // Enter this loop if the Functor has a init-join.
+      printf("Execute HasJoin in RangePolicy \n");
```
Comment (lucbv) [inline core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp:458]: I am not sure that you mean to leave this print statement in your code.

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp [hunk 2]
```diff
@@ -454,13 +455,16 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
...
       ParReduceSpecialize::execute_init_join(m_functor, m_policy, m_result_ptr,
                                              m_result_ptr_on_device);
     } else if constexpr (UseReducer) {
+      //    printf("Execute UseReducer in RangePolicy \n");
```
Comment (lucbv) [inline core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp:462]: Can probably be removed

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp [hunk 3]
```diff
@@ -454,13 +455,16 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
...
       // Enter this loop if the Functor is a reducer type.
       ParReduceSpecialize::execute_reducer(m_functor, m_policy, m_result_ptr,
                                            m_result_ptr_on_device);
     } else if constexpr (IsArray) {
+      //    printf("Execute IsArray in RangePolicy \n");
```
Comment (lucbv) [inline core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp:467]: Same as above

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp [hunk 4]
```diff
@@ -1204,12 +1204,15 @@ class ParallelReduce<FunctorType, Kokkos::TeamPolicy<Properties...>,
  public:
   void execute() const {
     if constexpr (HasJoin) {
+      printf("Execute HasJoin in TeamPolicy \n");
```
Comment (lucbv) [inline core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp:1207]: Same as above

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp [hunk 5]
```diff
@@ -1204,12 +1204,15 @@ class ParallelReduce<FunctorType, Kokkos::TeamPolicy<Properties...>,
...
       ParReduceSpecialize::execute_init_join(m_functor, m_policy, m_result_ptr,
                                              m_result_ptr_on_device);
     } else if constexpr (UseReducer) {
+      //    printf("Execute UseReducer in TeamPolicy \n");
```
Comment (lucbv) [inline core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp:1211]: Same as above

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp [hunk 6]
```diff
@@ -1204,12 +1204,15 @@ class ParallelReduce<FunctorType, Kokkos::TeamPolicy<Properties...>,
...
       ParReduceSpecialize::execute_reducer(m_functor, m_policy, m_result_ptr,
                                            m_result_ptr_on_device);
     } else if constexpr (IsArray) {
+      //    printf("Execute IsArray in TeamPolicy \n");
```
Comment (lucbv) [inline core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp:1215]: Same as above

**Review by lucbv — APPROVED** [review]
Overall this looks good to me but I would clean it up a little.

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp (hunk 1 above)
Comment (rgayatri23) [reply core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp]: Oh this was a mistake. I forgot to delete these after debugging 🤦 . Deleted them. Thanks @lucbv

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp [hunk 7]
```diff
@@ -168,7 +168,8 @@ struct ParallelReduceSpecialize<FunctorType, Kokkos::RangePolicy<PolicyArgs...>,
     const auto begin = p.begin();
     const auto end   = p.end();
 
-    ValueType result = ValueType();
+    ValueType result;
+    OpenMPTargetReducerWrapper<ReducerType>::init(result);
```
Comment (dalg24) [inline core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp:172]: The same initialization happens L179 after the `omp declare reduction` clause

Comment (rgayatri23) [reply core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp:172]: > The same initialization happens L179 after the `omp declare reduction` clause

Fixed it.

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp [hunk 8]
```diff
@@ -1023,32 +1023,28 @@ struct ParallelReduceSpecialize<FunctorType, TeamPolicyInternal<PolicyArgs...>,
...
       Kokkos::abort(
           "OpenMPTarget: Scratch memory is not supported in `parallel_reduce` "
           "over functors with init/join.");
     }
 
-    // Maximum active teams possible.
-    int max_active_teams = OpenMPTargetExec::MAX_ACTIVE_THREADS / team_size;
-    const auto nteams =
-        league_size < max_active_teams ? league_size : max_active_teams;
+    const auto nteams = league_size;
```
Comment (dalg24) [inline core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp:1047]: Did you mean to change this?

Comment (rgayatri23) [reply core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp:1047]: Yes the `nteams` should be equal to `league_size` . No need to include the `max_active_teams` in this case.

**Pull request merged**